src/ostrowski/rationals/basic.lean

```diff
@@ -0,0 +1,226 @@
+    },
+  }
+
+-- Deserves its place in matlib, as `monoid_with_zero_hom.map_inv`
+theorem Γ₀_map_inv {G₁ G₂: Type} [group_with_zero G₁] [group_with_zero G₂]
+  (φ: monoid_with_zero_hom G₁ G₂) (a: G₁) (a_ne_zero: a ≠ 0): φ a⁻¹ = (φ a)⁻¹ :=
+begin
+  apply eq_inv_of_mul_left_eq_one,
+  rw ← monoid_with_zero_hom.map_mul,
+  rw inv_mul_cancel a_ne_zero,
+  rw monoid_with_zero_hom.map_one,
+end
+
+lemma induction_on_primes [P: nat → Prop] (h₁: P 0) (h₂: P 1)
+  (h: ∀ p a: ℕ, prime p → P a → P (p * a)): ∀ n: ℕ, P n :=
+begin
+  intro n,
+  apply unique_factorization_monoid.induction_on_prime,
+  exact h₁,
+  {
+    intros n h,
+    rw nat.is_unit_iff.1 h,
+    exact h₂,
+  },
+  {
+    intros a p _ hp ha,
+    exact h p a hp ha,
+  },
+end
+
+-- Integer values of a morphism `φ` and its value on `-1` completely determines `φ`
+theorem mul_mor_eq_of_eq_on_pnat (φ₁ φ₂: monoid_with_zero_hom ℚ ℝ)
+  (same_on_neg_one: φ₁ (-1) = φ₂ (-1)) (same_on_nat: ∀ n: ℕ, φ₁ n = φ₂ n): (φ₁: ℚ → ℝ) = φ₂ :=
+begin
+  ext,
+  suffices : ∀ z : ℤ, φ₁ z = φ₂ z,
+  begin
+    induction x,
+    rw rat_mk_eq_div,
+    repeat { rw div_eq_mul_inv, },
+    have x_denom_ne_zero: (x_denom: ℚ) ≠ 0,
+    {
+      symmetry,
+      norm_cast,
+      exact (ne_of_lt x_pos),
+    },
+    rw monoid_with_zero_hom.map_mul φ₁,
+    rw Γ₀_map_inv φ₁ x_denom x_denom_ne_zero,
+    rw monoid_with_zero_hom.map_mul φ₂,
+    rw Γ₀_map_inv φ₂ x_denom x_denom_ne_zero,
+    rw this x_num,
+    have := this (↑ x_denom),
+    norm_cast at this,
+    rw this,
+  end,
+  intro x,
+  suffices : ∀ z : ℕ, φ₁ z = φ₂ z,
+  begin
+    induction x,
+    simp [this],
+    push_cast,
+    conv {
+      congr,
+      rw neg_eq_neg_one_mul,
+      skip,
+      rw neg_eq_neg_one_mul,
+      skip,
+    },
+    rw monoid_with_zero_hom.map_mul φ₁,
+    rw monoid_with_zero_hom.map_mul φ₂,
+    rw same_on_neg_one,
+    simp,
+    norm_cast,
+    exact this _,
+  end,
+  intro x,
+  cases x,
+  simp,
+  have := same_on_nat (nat.succ x_1),
+  simp [-add_comm] at this,
+  simp [-add_comm],
+  rw this,
+end
+
+lemma abs_val_equiv_of_equiv_on_primes (abv abv': ℚ → ℝ)
+  [habv: is_absolute_value abv] [habv': is_absolute_value abv']
+  (α: ℝ) (α_ne_zero: α ≠ 0)
+  (h: ∀ p: ℕ, prime p → (abv p) ^ α = abv' p): (λ r: ℚ, (abv r) ^ α) = abv' :=
+begin
+  have: ∀ n: ℕ, (abv n) ^ α = abv' n,
+  {
+    have inductive_step: ∀ q a: ℕ, prime q → (abv a) ^ α = abv' a
+      → (abv (q * a: ℕ)) ^ α = abv' ((q * a): ℕ),
+    {
+      intros q a q_prime a_norms_eq,
+      calc (abv (q * a: ℕ)) ^ α
+        = abv (q*a: ℕ) ^ α
+          : by { refl }
+        ... = ((abv q) * (abv a)) ^ α
+          : by { push_cast, rw abv_mul abv, }
+        ... = (abv q) ^ α * (abv a) ^ α
+          : by { rw real.mul_rpow (abv_nonneg abv q) (abv_nonneg abv a), }
+        ... = (abv' q) * (abv' a)
+          : by { rw [h q q_prime, a_norms_eq], }
+        ... = abv' (q * a)
+          : by { rw ← abv_mul abv', }
```

Comment:
you could merge the next step using rw_mod_cast

Reply:
Did not work :/
Used `rw ..., norm_cast` instead, won two lines.